### PR TITLE
Implement api retry on rate limits

### PR DESCRIPTION
**Pull request by afinkenstadt**
Implements #46

### CODE REVIEW

File: hipchat/hipchat.go [hunk 1]
```diff
@@ -275,6 +325,58 @@ func (c *Client) Do(req *http.Request, v interface{}) (*http.Response, error) {
...
+	currentTry := 0
+
+	for willContinue(currentTry, policy) {
+		currentTry = currentTry + 1
+		resp, err := c.client.Do(req)
+		if err != nil {
+			return nil, err
+		}
+		c.captureRateLimits(resp)
+		if code := resp.StatusCode; code == http.StatusTooManyRequests {
```
Comment (tbruyelle) [inline hipchat/hipchat.go]:
```go
if resp.StatusCode ==  http.StatusTooManyRequests {
```
You don't reuse the code so the var is useless I think.

File: hipchat/hipchat.go [hunk 2]
```diff
@@ -275,6 +325,58 @@ func (c *Client) Do(req *http.Request, v interface{}) (*http.Response, error) {
...
+	}
+	return nil, fmt.Errorf("max retries exceeded (%d)", policy.MaxRetries)
+}
+
+func willContinue(currentTry int, policy RetryPolicy) bool {
+	return currentTry <= policy.MaxRetries
+}
+
+func doDelay(currentTry int, policy RetryPolicy) {
+	jitter := time.Duration(rand.Int63n(2*int64(policy.JitterDelay))) - policy.JitterBias
```
Comment (tbruyelle) [inline hipchat/hipchat.go]: Why do you randomize the delay ? What's the purpose of `JitterBias` ?

File: hipchat/hipchat.go [hunk 3]
```diff
@@ -275,6 +325,58 @@ func (c *Client) Do(req *http.Request, v interface{}) (*http.Response, error) {
...
+	time.Sleep(time.Duration(linearDelay))
+}
+
+func setIfPresent(src string, dest *int) {
+	if len(src) > 0 {
+		v, err := strconv.Atoi(src)
+		if err != nil {
+			*dest = v
+		}
+	}
```
Comment (tbruyelle) [inline hipchat/hipchat.go]: When the header is not present, `Client` will keep the RateLimit/FloodControl values of the previous response with that headers. Is this expected ?

File: hipchat/hipchat.go [hunk 4]
```diff
@@ -275,6 +325,58 @@ func (c *Client) Do(req *http.Request, v interface{}) (*http.Response, error) {
...
+	return nil, fmt.Errorf("max retries exceeded (%d)", policy.MaxRetries)
+}
+
+func willContinue(currentTry int, policy RetryPolicy) bool {
+	return currentTry <= policy.MaxRetries
+}
+
+func doDelay(currentTry int, policy RetryPolicy) {
```
Comment (tbruyelle) [inline hipchat/hipchat.go]: `doDelay()` name creates ambiguity with request scoped functions like `Do` or `doWithRetryPolicy`. That's maybe dumb but when I first read this I thought the function triggered the request. Maybe you should rename it to `sleep()` ?

File: hipchat/hipchat.go [hunk 5]
```diff
@@ -100,6 +116,35 @@ var AuthTest = false
...
+
+// NoRateLimitRetryPolicy defines the "never retry an API call" policy's values.
+var NoRateLimitRetryPolicy = RetryPolicy{0, 1 * time.Second, 1 * time.Second, 500 * time.Millisecond, 0 * time.Millisecond}
+
+// DefaultRateLimitRetryPolicy defines the "up to 300 times, 1 second apart, randomly adding an additional up-to-500 milliseconds of delay" policy.
+var DefaultRateLimitRetryPolicy = RetryPolicy{300, 1 * time.Second, 1 * time.Second, 500 * time.Millisecond, 0 * time.Millisecond}
+
+// RateLimitRetryPolicy can be set to a custom RetryPolicy's values,
+// or to one of the two predefined ones: NoRateLimitRetryPolicy or DefaultRateLimitRetryPolicy
+var RateLimitRetryPolicy = &DefaultRateLimitRetryPolicy
```
Comment (afinkenstadt) [inline hipchat/hipchat.go]: This implementation was inspired by some other work where the `nil` `RetryPolicy` was effectively the `NO` switch.    I was also prematurely optimizing against memory copies of an inexpensive struct (nanoseconds) in comparison to network delay (milliseconds).  

Will address.

File: hipchat/hipchat.go (hunk 1 above)
Comment (afinkenstadt) [inline hipchat/hipchat.go]: Addressed in 4d20516

File: hipchat/hipchat.go (hunk 4 above)
Comment (afinkenstadt) [inline hipchat/hipchat.go]: Addressed in 3c2b1e0
Also fixed a leak in `resp.Body` when it remains unused.

File: hipchat/hipchat.go (hunk 3 above)
Comment (afinkenstadt) [inline hipchat/hipchat.go]: This is by design.  Floodcontrol is only updated when a room or user notification/message is sent.  It has a 30-per-minute limit, which is orthogonal to the API rate limit value of 100 (for user tokens) or 500 (for app tokens).

Will document in 175d404.

File: hipchat/hipchat.go (hunk 2 above)
Comment (afinkenstadt) [inline hipchat/hipchat.go]: Addressed in 3321cfc

**Comment by afinkenstadt** [comment]
In 93e7a7f I also reformatted the godoc comments to be more customary.

**Pull request merged**